Google.Api.Gax.Grpc/Gcp/GcpCallInvokerWithRefresh.cs

```diff
@@ -0,0 +1,539 @@
+﻿/*
+ * Copyright 2018 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Protobuf;
+using Google.Protobuf.Collections;
+using Grpc.Core;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc.Gcp
+{
+    /// <summary>
+    /// Call invoker which can fan calls out to multiple underlying channels
+    /// based on request properties.
+    /// </summary>
+    public sealed class GcpCallInvokerWithRefresh : CallInvoker
```

Comment:
I'd really prefer us to avoid the duplication here. I was hoping we could use the existing GcpCallInvokerPool and GcpCallInvoker classes, either making `ChannelRef` itself handle the reference or (more likely) making GcpCallInvoker do so.

Reply:
Noted Jon, I will raise another draft pr for that option. 
Thanks for providing your feedback,